### ISSUE TITLE
Customize SSL cert renewing interval to avoid reloading nginx every hour

### ISSUE THREAD

**Issue by niklasdahlheimer** [jc21/nginx-proxy-manager#677]
**Is your feature request related to a problem? Please describe.**
First of all: great tool! Setting up reverse poxies is fun with this one :)  

According to the logs, npm is checking for expired certificates every hour:
```
[10/26/2020] [1:26:22 AM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[10/26/2020] [1:26:24 AM] [Nginx    ] › ℹ  info      Reloading Nginx
[10/26/2020] [1:26:24 AM] [SSL      ] › ℹ  info      Renew Complete
[10/26/2020] [2:26:22 AM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[10/26/2020] [2:26:24 AM] [Nginx    ] › ℹ  info      Reloading Nginx
[10/26/2020] [2:26:24 AM] [SSL      ] › ℹ  info      Renew Complete
[10/26/2020] [3:26:22 AM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[10/26/2020] [3:26:23 AM] [Nginx    ] › ℹ  info      Reloading Nginx
[10/26/2020] [3:26:23 AM] [SSL      ] › ℹ  info      Renew Complete
[...]
```
I think reloading nginx every hour could lead to problems in the business times of our web services (e.g. downloads buffered by nginx ..)

**Is there a way to customize the cron job which triggers the cert renewing every hour to start the renewing only once a day or so?** 

**Describe the solution you'd like**
some way (UI or production.config) to customize the renawal interval

**Comment by tophee** [comment]
I don't quite understand why the renewing interval is set to such a short period in the first place. I think, letsencrypt certs are usually valid for three months or so. Why do 2160 checks during that period?

**Comment by niklasdahlheimer** [comment]
probably it's just changing
`intervalTimeout:    1000 * 60 * 60, // 1 hour`
to
`intervalTimeout:    (process.env.CERTIFICATE_RENEWAL_INTERVAL_IN_SECONDS * 1000) || (1000 * 60 * 60), // 1 hour`

in https://github.com/jc21/nginx-proxy-manager/blob/4f10d129c20cc82494b95cc94b97f859dbd4b54d/backend/internal/certificate.js#L24

Than the environment variable  `CERTIFICATE_RENEWAL_INTERVAL_IN_SECONDS ` could be used to set the preferred interval.

Sadly I'm not able to setup a working build environment to test it to create a pull request :/

**Comment by niklasdahlheimer** [comment]
@chaptergy I agree with you that certbot will only contact letsEncrypt if a certificate really expires within the next 30 days:
_"[...]Since renew only renews certificates that are near expiry it can be run as frequently as you want - since it will usually take no action.[...]"_
https://certbot.eff.org/docs/using.html?highlight=renew#renewing-certificates

**But** my concerns where more about the reload of nginx, which, I think, is triggered every time the task is run, so every hour. See
https://github.com/jc21/nginx-proxy-manager/blob/5c679084605a79d98661a65eb272a06fc8cb5669/backend/internal/certificate.js#L55

I'm not so much into nginx but couldn't this might cause running connections to "hickup" or result in a denial of service if you do a request just in this reload interval? 
Either way, this hourly reload is not nescessary and I think it's easy to avoid it by implementing a environment variable, just like I suggested above.

**Comment by chaptergy** [comment]
It seems the original intention was to not run certbot at boot since you can't execute other certbot commands while it is running. And to make sure certbot runs even if you restart frequently, the hourly interval was chosen.

**Comment by tophee** [comment]
Thanks for the explanation. At least it makes some sense now. But the same could also be achieved by running certbot exactly once an hour after boot and thereafter once every couple of weeks or so.

**Comment by marcb1387** [comment]
I'm having this same thing occur and getting 522 error on my pages around the time when it reloads nginx. All of my certs are months off from expiration. If this could be scheduled during off time one a day it would be great. 

[2/7/2022] [9:05:39 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[2/7/2022] [9:05:41 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[2/7/2022] [9:05:41 PM] [SSL      ] › ℹ  info      Renew Complete

**Comment by RafaelSchridi** [comment]
Since it renews when expiry is <30 days I see absolutely no reason to run this so often, should be at the minimum a 1 day interval.

**Comment by LucidityCrash** [comment]
>  Now the port 80 tcp is opened in every moment since we don't know when the renew will happen.

Is it really opening port 80 though ... I don't claim to be an expert but it doesn't make sense to me for it to be attaching to port 80 if there isn't a certificate to renew ... that said I use the cloudflare dns method so it never opens port 80 and I can't test either

however from the certbot docs : 
```certbot renew

This command attempts to renew any previously-obtained certificates that expire in less than 30 days. 
The same plugin and options that were used at the time the certificate was originally issued will be used
for the renewal attempt, unless you specify other plugins or options. Unlike certonly, renew acts on
multiple certificates and always takes into account whether each one is near expiry. Because of this, 
renew is suitable (and designed) for automated use, to allow your system to automatically renew each
certificate when appropriate. Since renew only renews certificates that are near expiry it can be run as
frequently as you want - since it will usually take no action.```

Note this bit  **it can be run as frequently as you want - since it will usually take no action**  It really does seem to suggest that this is not doing what you seem to believe it is doing.

**Comment by frabnet** [comment]
I don't want to criticize, just clarify:  I assure you that the port 80 is always opened and always responsive, even when a renew is not needed:
```
frab@test:~$ curl --head http://website
HTTP/1.1 301 Moved Permanently
Server: openresty
Date: Wed, 04 May 2022 07:37:15 GMT
Content-Type: text/html
Content-Length: 166
Connection: keep-alive
Location: https://website/
```
With a scheduled renew timing, it could be closed by the user firewall (it can be a hardware firewall or iptables rule in the container).

**Comment by LucidityCrash** [comment]
I'm really confused here, I certainly don't think you are critizing, I just think we seem to have a missunderstanding :)

The points I've made are that 

1. NPM does an nginx reload which wont interrupt a existing connection
2. certbot only actually actually does the renew process when there is something to renew.

Looking at the cert code in NPM : 
Initial cert request - https://github.com/NginxProxyManager/nginx-proxy-manager/blob/14b889a85f2f8af9a13ed6122f5a0a91d64ecc36/backend/internal/certificate.js#L835
and renewal - https://github.com/NginxProxyManager/nginx-proxy-manager/blob/14b889a85f2f8af9a13ed6122f5a0a91d64ecc36/backend/internal/certificate.js#L42
(renews are done with the same authenticator as the initial request)

I can see that if you do not use the DNS challenge method then it will use certbot with the webroot auth method with certonly ... this means that the web service, in this case nginx, is not restarted. certbot adds files to specific paths for the auth process  - https://eff-certbot.readthedocs.io/en/stable/using.html?highlight=renew#webroot

So in short even if you don't use dns challenge for LE your sites should never be unavailable (slight caveat on that - https://github.com/NginxProxyManager/nginx-proxy-manager/blob/14b889a85f2f8af9a13ed6122f5a0a91d64ecc36/backend/internal/certificate.js#L131 if you create a new certificate that has a domain name in that is already in another cert then the site using that other cert gets disabled while the new cert is being created)

I'm rereading what you are saying and I think I understand what you are saying Port 80 is always listening and you'd like to be able to disable port 80 completely (except for cert renewals) ?

If that is the case then I don't think this is done because of LE ... this is simply the way NPM sets up its server config files you kinda have have to have it listening on port 80 else how would you serve up sites that you don't want SSLd ?  and if you enable "force SSL" on a site then extra config is injected for that site that puts in that redirect you posted.

OK you could argue that in this day and age why would you not SSL everything but this is a generic Proxy, you have to account for all use cases and as you pointed out yourself having port 80 open is no real security threat - though changing the default to something other than the Congratulations and 404 page is probably a good idea as both of those give hints as to what is running ... the custom page option does too as it sets the server header :(  But that doesn't really increase the attack surface just gives them a little more info that will save them all of 10 seconds.

**Comment by waytosahil** [comment]
I see it is mentioned in above thread that request is sent to Let's Encrypt only if the cert is due for renewal and renewal check done locally (my understanding from above thread). 

One observation from my setup, I have 19 certs installed  and I can see in the last 24 hrs ~700 requests were made by NPM to lencr.org ,  which are too many request in my opinion for certs with expiry time > 2 Months. May be if we can have an option to manage it would really help optimize things.

**Comment by pShota** [comment]
(bump) may I ask any update on this?

**Comment by kirby88bit** [comment]
I would like to add to this thread (know it's old but it's still open) this is causing a hiccup with using streams. I am running minecraft servers and this is causing external connections to drop during each reload.

**Comment by sercanyaldiz** [comment]
Any update on this?

My certificate is expires 11th July 2023, 2:45 am

Why does NPM renew my SSL cert everyh 1 hour?
```
[5/4/2023] [3:19:54 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[5/4/2023] [3:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [3:19:54 PM] [SSL      ] › ℹ  info      Renew Complete
[5/4/2023] [4:19:54 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[5/4/2023] [4:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [4:19:54 PM] [SSL      ] › ℹ  info      Renew Complete
[5/4/2023] [5:19:54 PM] [IP Ranges] › ℹ  info      Fetching IP Ranges from online services...
[5/4/2023] [5:19:54 PM] [IP Ranges] › ℹ  info      Fetching https://ip-ranges.amazonaws.com/ip-ranges.json
[5/4/2023] [5:19:54 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[5/4/2023] [5:19:54 PM] [IP Ranges] › ℹ  info      Fetching https://www.cloudflare.com/ips-v4
[5/4/2023] [5:19:54 PM] [IP Ranges] › ℹ  info      Fetching https://www.cloudflare.com/ips-v6
[5/4/2023] [5:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [5:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [5:19:54 PM] [SSL      ] › ℹ  info      Renew Complete
[5/4/2023] [6:19:54 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[5/4/2023] [6:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [6:19:54 PM] [SSL      ] › ℹ  info      Renew Complete
[5/4/2023] [7:19:54 PM] [SSL      ] › ℹ  info      Renewing SSL certs close to expiry...
[5/4/2023] [7:19:54 PM] [Nginx    ] › ℹ  info      Reloading Nginx
[5/4/2023] [7:19:54 PM] [SSL      ] › ℹ  info      Renew Complete
```

**Comment by JOHLC** [comment]
I also agree that a configurable renewal period would be a good feature. I have multiple instances requesting certs, every hour seems excessive.

**Comment by MyCleverGirl** [comment]
im having the same issues, every hour all domains go offline untill its done and reloads, i tried to change the (1000 * 60 * 60) to 30 days and i get a sqlite error in docker logs and system breaks till i re- add (1000 * 60 * 60). im thinking of looking for alteritive software, its a great program but this issue makes me wanna go elese were.

**Comment by fullmetalbrackets** [comment]
This issue has been open since 2020, will it ever be resolved or even acknowledged by the maintainers? I don't even use SSL certificates, I only proxy my Docker containers to `*.home.arpa` locally via http, and it still does this renewal every hour on the hour. There aren't even any SSL certificates to renew!

**Comment by HomemadeAdvanced** [comment]
Nginx Proxy Manager V3 will be released. It is in development since two years. So it is obvious that the developers invest not that much afford into the NPM V2 anymore.

**Comment by HomemadeAdvanced** [comment]
Is still an issue.

**Comment by jjmerri** [comment]
why does reloading cause unavailability? I wouldn't care about this hourly check if it wasn't causing outages.

**Comment by kramttocs** [comment]
So if it's checking every hour, I only need to open port 80 on my router, let's say as an example, for a period of 2hrs once a week, right? 

Just thinking it'd be nice to not have port 80 open on the router if it isn't needed 24/7.

**Comment by BuckinghamIO** [comment]
5 years and no update on something that really is so very simple...

This is causing me issues now as the reload knocks off the home assistant app it seems.

**Comment by kramttocs** [comment]
> 5 years and no update on something that really is so very simple...
> 
> This is causing me issues now as the reload knocks off the home assistant app it seems.

Are you on the latest? Pretty sure the reload part was addressed with: https://github.com/NginxProxyManager/nginx-proxy-manager/commit/1be87f48c156fdbbe981d3d5ff73758f2718c93c

**Comment by dragonfly4** [comment]
This is still running every hour and causing my Docker container to hang. Certificate expires on September 15, 2025.

https://github.com/NginxProxyManager/nginx-proxy-manager/blob/f327c1e825f3049309edbdbe3c0ebd423b51211d/backend/internal/certificate.js#L34

[6/23/2025] [11:43:47 AM] [SSL      ] › ℹ  info      Renewing SSL certs expiring within 30 days ...
[6/23/2025] [11:43:47 AM] [SSL      ] › ℹ  info      Completed SSL cert renew process

**Comment by dominic-preap** [comment]
Is there any way to bypass on that?

**Comment by choddles** [comment]
This is still an issue
